### PR TITLE
Add some helpful functions for getting attachments

### PR DESCRIPTION
**Pull request by JustRed23**
[contributing]: https://jda.wiki/contributing/contributing/

## Pull Request Etiquette

- [x] I have checked the PRs for upcoming features/bug fixes.
- [x] I have read the [contributing guidelines][contributing].

### Changes

- [x] Internal code
- [ ] Library interface (affecting end-user code) 
- [x] Documentation
- [ ] Other: \_____ 

Closes Issue: NaN

## Description

This PR adds some extra functions to make getting specific attachments easier, it also includes a function to check if the message actually has attachments.

PS: I hope the documentation is good enough ;)

### CODE REVIEW

File: src/main/java/net/dv8tion/jda/api/entities/Message.java [hunk 1]
```diff
@@ -555,17 +555,60 @@ default String getApplicationId()
...
      */
     @Nonnull
     List<Attachment> getAttachments();
 
+    /**
+     * Returns {@code true} if this message has any attachments with the provided extension.
+     *
+     * <p><b>Requires {@link net.dv8tion.jda.api.requests.GatewayIntent#MESSAGE_CONTENT GatewayIntent.MESSAGE_CONTENT}</b>
+     *
+     * @param extension The extension to filter by (ex. "png")
```
Comment (freya022) [inline src/main/java/net/dv8tion/jda/api/entities/Message.java:588]:
```suggestion
     * @param extension  The extension to filter by (ex. "png")
```

File: src/main/java/net/dv8tion/jda/api/entities/Message.java [hunk 2]
```diff
@@ -555,17 +555,60 @@ default String getApplicationId()
...
+     *
+     * @return True, if this message has an attachment with the provided extension.
+     *
+     * @see    #getAttachmentsByExtension(String)
+     */
+    boolean hasAttachmentsByExtension(@Nonnull String extension);
+
+    /**
+     * An immutable list of {@link net.dv8tion.jda.api.entities.Message.Attachment Attachments} that are attached to this message, filtered by file extension.
+     * <br>Most likely this will only ever be 1 {@link net.dv8tion.jda.api.entities.Message.Attachment Attachment} at most.
```
Comment (freya022) [inline src/main/java/net/dv8tion/jda/api/entities/Message.java:598]: Why would there be at most 1 attachment with the same extension ?

File: src/main/java/net/dv8tion/jda/api/entities/Message.java [hunk 3]
```diff
@@ -555,17 +555,60 @@ default String getApplicationId()
...
+     *
+     * <p><b>Requires {@link net.dv8tion.jda.api.requests.GatewayIntent#MESSAGE_CONTENT GatewayIntent.MESSAGE_CONTENT}</b>
+     *
+     * @param extension The extension to filter by (ex. "png")
```
Comment (freya022) [inline src/main/java/net/dv8tion/jda/api/entities/Message.java:602]:
```suggestion
     * @param extension  The extension to filter by (ex. "png")
```

File: src/main/java/net/dv8tion/jda/internal/entities/ReceivedMessage.java [hunk 4]
```diff
@@ -460,6 +461,12 @@ public Guild getGuild()
         return getGuildChannel().getGuild();
     }
 
+    @Override
+    public boolean hasAttachments()
+    {
+        return getAttachments().size() > 0;
```
Comment (freya022) [inline src/main/java/net/dv8tion/jda/internal/entities/ReceivedMessage.java:467]:
```suggestion
        return !getAttachments().isEmpty();
```

File: src/main/java/net/dv8tion/jda/internal/entities/ReceivedMessage.java [hunk 5]
```diff
@@ -468,6 +475,24 @@ public List<Attachment> getAttachments()
         return attachments;
     }
 
+    @Override
+    public boolean hasAttachmentsByExtension(@Nonnull String extension)
+    {
+        Checks.notEmpty(extension, "Extension");
+        return getAttachmentsByExtension(extension).size() > 0;
```
Comment (freya022) [inline src/main/java/net/dv8tion/jda/internal/entities/ReceivedMessage.java:482]:
```suggestion
        return !getAttachmentsByExtension(extension).isEmpty();
```

File: src/main/java/net/dv8tion/jda/internal/entities/ReceivedMessage.java [hunk 6]
```diff
@@ -468,6 +475,24 @@ public List<Attachment> getAttachments()
...
+
+    @Nonnull
+    @Override
+    public List<Attachment> getAttachmentsByExtension(@Nonnull String extension)
+    {
+        Checks.notEmpty(extension, "Extension");
+        List<Attachment> results = getAttachments().stream() //No need to check intents here, we already did in getAttachments()
+                .filter(a -> Objects.equals(a.getFileExtension(), extension))
+                .collect(Collectors.toList());
+        return Collections.unmodifiableList(results);
```
Comment (freya022) [inline src/main/java/net/dv8tion/jda/internal/entities/ReceivedMessage.java:493]:
```suggestion
        return getAttachments().stream() //No need to check intents here, we already did in getAttachments()
                .filter(a -> Objects.equals(a.getFileExtension(), extension))
                .collect(Helpers.toUnmodifiableList());
```

**Review by freya022 — CHANGES_REQUESTED** [review]

File: src/main/java/net/dv8tion/jda/api/entities/Message.java (hunk 2 above)
Comment (JustRed23) [reply src/main/java/net/dv8tion/jda/api/entities/Message.java:598]: This warning was included in getAttachments() so I thought I'll just add it to this as well

Comment (freya022) [reply src/main/java/net/dv8tion/jda/api/entities/Message.java:598]: Yea I guess, but desktop users aren't limited to 1 attachment anymore, might as well remove it imo

Comment (JustRed23) [reply src/main/java/net/dv8tion/jda/api/entities/Message.java:598]: Sure

**Comment by MinnDevelopment** [comment]
Thank you, but we generally do not implement methods that can be replaced by trivial stream operations. We do not want to bloat the interface with a lot of unnecessary helper methods.

**Comment by JustRed23** [comment]
> Thank you, but we generally do not implement methods that can be replaced by trivial stream operations. We do not want to bloat the interface with a lot of unnecessary helper methods.

What if i make an event for it? Like an AttachmentReceivedEvent

**Comment by Chew** [comment]
Events are generally mapped to Discord events, and there is no such AttachmentReceivedEvent, you're better off just making a util class in your project

**Comment by MinnDevelopment** [comment]
That would be a very unnecessary addition since it would only remove a single boolean check in user code. You trade a single line of code for:

- An extra function in ListenerAdapter (complexity and memory overhead)
- An extra class to load at runtime (CPU time and memory overhead)
- Cognitive complexity (Why does this event exist? Does that mean normal message received has no attachment support? When do you receive this?)

Instead, a user can simply use a guard statement:

```java
if (event.getMessage().getAttachments().isEmpty()) return;
```

**Pull request closed without merging**